pallets/xyk-liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,927 @@
+			});
+
+			Ok(())
+		}
+
+		/// Claim rewards from liq. mining for deposit represented by `nft_id`.
+		///
+		/// This function calculate user rewards from liq. mining and transfer rewards to `origin`
+		/// account. Claiming in the same period is allowed only once.
+		///
+		/// Parameters:
+		/// - `origin`: account owner of deposit(nft).
+		/// - `deposit_id`: nft id representing deposit in the yield farm.
+		/// - `yield_farm_id`: yield farm identifier to claim rewards from.
+		///
+		/// Emits `RewardClaimed` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::claim_rewards())]
+		pub fn claim_rewards(
+			origin: OriginFor<T>,
+			deposit_id: DepositId,
+			yield_farm_id: YieldFarmId,
+		) -> DispatchResult {
+			let owner = Self::ensure_nft_owner(origin, deposit_id)?;
+
+			let fail_on_double_claim = true;
+			let (global_farm_id, reward_currency, claimed, _) = T::LiquidityMiningHandler::claim_rewards(
+				owner.clone(),
+				deposit_id,
+				yield_farm_id,
+				fail_on_double_claim,
+			)?;
+
+			ensure!(!claimed.is_zero(), Error::<T>::ZeroClaimedRewards);
+
+			Self::deposit_event(Event::RewardClaimed {
+				global_farm_id,
+				yield_farm_id,
+				who: owner,
+				claimed,
+				reward_currency,
+				deposit_id,
+			});
+
+			Ok(())
+		}
+
+		/// Withdraw LP shares from liq. mining. with reward claiming if possible.
+		///
+		/// Cases for transfer LP shares and claimed rewards:
+		///
+		/// * yield farm is active(yield farm is not stopped) - claim and transfer rewards(if it
+		/// wasn't claimed in this period) and transfer LP shares.
+		/// * liq. mining is stopped - claim and transfer rewards(if it
+		/// wasn't claimed in this period) and transfer LP shares.
+		/// * yield farm was destroyed - only LP shares will be transferred.
+		/// * farm was destroyed - only LP shares will be transferred.
+		/// * SPECIAL CASE: AMM pool does not exist - claim may happen if yield farm is still active, LP
+		/// shares will not be transferred.
+		///
+		/// User's unclaimable rewards will be transferred back to global farm's account.
+		///
+		/// Parameters:
+		/// - `origin`: account owner of deposit(nft).
+		/// - `deposit_id`: nft id representing deposit in the yield farm.
+		/// - `yield_farm_id`: yield farm identifier to dithdraw shares from.
+		/// - `asset_pair`: asset pair identifying yield farm in global farm.
+		///
+		/// Emits:
+		/// * `RewardClaimed` if claim happen
+		/// * `SharesWithdrawn` event when successful
+		#[pallet::weight(<T as Config>::WeightInfo::withdraw_shares())]
+		pub fn withdraw_shares(
+			origin: OriginFor<T>,
+			deposit_id: DepositId,
+			yield_farm_id: YieldFarmId,
+			asset_pair: AssetPair,
+		) -> DispatchResult {
+			let owner = Self::ensure_nft_owner(origin, deposit_id)?;
+			let amm_pool_id = Self::ensure_amm(asset_pair)?;
+
+			let global_farm_id = T::LiquidityMiningHandler::get_global_farm_id(deposit_id, yield_farm_id)
+				.ok_or(Error::<T>::DepositDataNotFound)?;
```

Comment:
Is this correct error ?

Reply:
Yes, this error is ok, but I'm open to suggestions for better name. 
Basically `get_global_farm_id()` can return `None` in 2 cases:
1-th - deposit doesn't exists - this case should never happen because NFT must exists at this point (`ensure_nft_owner()`)
2-nd - deposit does exists but there are not data for `yield_farm_id` in the deposit.